retry/retrier.go

```diff
@@ -0,0 +1,209 @@
+	if r.maxAttempts == 0 && !r.forever {
+		panic("retriers must either run forever, or have a maximum attempt count")
+	}
+
+	if r.maxAttempts < 0 {
+		panic("retriers must have a positive max attempt count")
+	}
+
+	oldJitter := r.jitter
+	r.jitter = false // Temporarily turn off jitter while we check if the interval is 0
+	if r.forever && r.strategyType == constantStrategy && r.NextInterval() == 0 {
+		panic("retriers using the constant strategy that run forever must have an interval")
+	}
+	r.jitter = oldJitter // and now set it back to what it was previously
+
+	return r
+}
+
+func (r *Retrier) calculateJitter() time.Duration {
+	if r.jitter {
+		return time.Duration(random.Float32()) * jitterInterval
+	}
+
+	return 0
+}
+
+// MarkAttempt increments the attempt count for the retrier. This affects ShouldGiveUp, and also affects the retry interval
+// for Exponential retry strategy
+func (r *Retrier) MarkAttempt() {
+	r.attemptCount += 1
+	r.lastAttemptAt = time.Now()
+}
+
+// Break causes the Retrier to stop retrying after it completes the next retry cycle
+func (r *Retrier) Break() {
+	r.breakNext = true
+}
+
+// ShouldGiveUp returns whether the retrier should stop trying do do the thing it's been asked to do
+// It returns true if the retry count is greater than r.maxAttempts, or if r.Break() has been called
+// It returns false if the retrier is supposed to try forever
+func (r *Retrier) ShouldGiveUp() bool {
+	if r.breakNext {
+		return true
+	}
+
+	if r.forever {
+		return false
+	}
+
+	return r.attemptCount > r.maxAttempts
+}
+
+// NextInterval returns the next interval that the retrier will use. Behind the scenes, it calls the function generated
+// by either retrier's strategy
+func (r *Retrier) NextInterval() time.Duration {
+	return r.intervalCalculator(r)
+}
+
+func (r *Retrier) String() string {
+	str := fmt.Sprintf("Attempt %d/", r.attemptCount)
+
+	if r.forever {
+		str = str + "∞"
+	} else {
+		str = str + fmt.Sprintf("%d", r.maxAttempts)
+	}
+
+	nextInterval := r.NextInterval()
+	if nextInterval > 0 {
+		str = str + fmt.Sprintf(" Retrying in %s", nextInterval-time.Since(r.lastAttemptAt))
+	} else {
+		str = str + " Retrying immediately"
+	}
+
+	return str
+}
+
+// Do is the core loop of a Retrier. It defines the operation that the Retrier will attempt to perform, retrying it if necessary
+// Calling retrier.Do(someFunc) will cause the Retrier to attempt to call the function, and if it returns an error,
+// retry it using the settings provided to it.
+func (r *Retrier) Do(callback func(*Retrier) error) error {
+	var err error
+	for {
+		r.MarkAttempt()
+
+		// If the last callback called r.Break(), or if we've hit our call limit, bail out and return the last error we got (or nil)
+		if r.ShouldGiveUp() {
+			return err
+		}
+
+		// Perform the action the user has requested we retry
+		err = callback(r)
+		if err == nil {
+			return nil
+		}
+
+		time.Sleep(r.NextInterval())
```

Comment:
This loop is not in the order I expected it to be. I expected:

1. Call the callback
2. Return if success
4. MarkAttempt (because in my head it's a count of failures so it happens after the call, but it could be earlier)
3. Check ShouldGiveUp
5. Call NextInterval and sleep

I think the main differences are:

1. If we're going to give up, we should do so immediately, not after sleeping.
2. If I set MaxAttempts to 1, it should make one request but the current code won't make any requests.

Reply:
yep, these exact points came up in the mob review session we did the other day. the main reason i did it like this was to maintain compatibility with the old retrier, which behaved in the same way as laid out in the code. there's one test that fails if we shuffle around the order, but IMO i'd rather change some tests such that our code is semantically correct than make code that does weird stuff so that i don't have to change tests

Reply:
okay, so i finally got around to fixing this today, but in the latest push, we get the best of both worlds - clear logic in the retry loop, and we didn't need to change any of the existing tests :)